### PR TITLE
pyBabyAFQ

### PR DESCRIPTION
**Pull request by bloomdt-uw**
#487 

Based on Mareike's recommendation, pausing work on dHCP dataset and generally abandoning approach to use FA. Instead will focus on replicating MATLAB AFQ results with Kalanit longitudinal data.

### [Kalanit longitudinal data](https://s3.console.aws.amazon.com/s3/buckets/dhcp-afq/Kalanit_Longitudinal_BabyData/)

Processed using [MATLAB BabyAFQ](https://github.com/yeatmanlab/AFQ/tree/babyAFQ)

Goal is to reproduce results in: 
- [Gold standard](https://s3.console.aws.amazon.com/s3/buckets/dhcp-afq/Kalanit_Longitudinal_BabyData/dti94trilin/fibers/afq/manual_tracts/)

- [MATLAB AFQ](https://s3.console.aws.amazon.com/s3/buckets/dhcp-afq/Kalanit_Longitudinal_BabyData/dti94trilin/fibers/afq/)

- [ ] use existing tractography data and feed into pyAFQ segmentation

    - [ ] convert dataset to BIDS compliant
    - [ ] may need to use [PyAFQ scripts](https://github.com/36000/pyAFQ-scripts) to convert MATLAB AFQ files
    - [ ] verify can convert `mif` files to `nii`
    - [ ] possible to use two step process and segment using CLI?

- [ ] if successful, then consider tractography using ACT within pyAFQ

### [dHCP data](https://s3.console.aws.amazon.com/s3/buckets/dhcp-afq/dhcp_dmri_pipeline/)

[Single Subject data](https://s3.console.aws.amazon.com/s3/buckets/dhcp-afq/dhcp_dmri_pipeline/sub-CC00060XX03/ses-12501/dwi/)

- [ ] if Kalanit pipeline sucessful, continue with dHCP

Revisit BIDS conversion

- using default dHCP dataset_description.json generates BIDS validation error.

   dHCP uses BIDS 1.0.2 ([dHCP Data Release Notes](https://dhcp-afq.s3.us-west-2.amazonaws.com/dhcp_dmri_pipeline/dHCP_Data_organization2019.pdf))

   - [ ] converter from BIDS 1.0.2 to BIDS 1.4.0

   - [ ] add BIDSVersion to original file

   - [ ] document and continue to use workaround: replace `"PipelineDescription.Name"` with `"PipelineDescription": { "Name"

- Error processing `sub-CC#####XX##_ses-#####_desc-preproc_dwi.json` file when present. Currently just excluded file

   - [ ] investigate error

- when making BIDS compliant pyAFQ required data in AFQ_data `derivatives` folder

   - [ ] add to documentation

- pyAFQ does not read `nii` files must be `nii.gz`, see #151, therefore needed to gzip nii files.

- [ ] Convert script to Jupyter notebook with [CloudKnot](https://github.com/nrdg/cloudknot) (see #533)

- [ ] if pyAFQ tractography doesn't work, consider two-step pipeline: use qsiprep/mrtrix for tractography then pyAFQ segmentation

- [ ] add segmentation end point filtering

- [ ] dependency: custom ROI monkey patch can be removed once #519 merged

### debugging FA tractography

However if decide to revisit dHCP FA:

Parameters explored thus far (all without endpoint filtering):

- default

- b0 1000

- b0 1000 with FA seed threshold 0.1

- b0 1000 with ROIMask

- b0 1000 with BrainMask and ROIMask

- b0 1000 with BrainMask, ROIMask, and  FA stop threshold 0.1

In all cases some bundles were entirely missing and tractography was highly irregular.

### CODE REVIEW

File: examples/plot_baby_afq.py [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+    pediatric_templates['ARC_roi2_L'] = pediatric_templates['SLFt_roi2_L']
+    pediatric_templates['ARC_roi2_R'] = pediatric_templates['SLFt_roi2_R']
+    pediatric_templates['ARC_roi3_L'] = pediatric_templates['SLFt_roi3_L']
+    pediatric_templates['ARC_roi3_R'] = pediatric_templates['SLFt_roi3_R']
+
+    # For the middle longitudinal fasciculus (MdLF) reuse ILF ROI
+    pediatric_templates['MdLF_roi2_L'] = pediatric_templates['ILF_roi1_L']
+    pediatric_templates['MdLF_roi2_R'] = pediatric_templates['ILF_roi1_R']
+
+    return pediatric_templates
```
Comment (arokem) [inline examples/plot_baby_afq.py:215]: All of this stuff should move into the `data` module.

File: examples/plot_baby_afq.py [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+                    'ROIs': [pediatric_templates[name + '_roi1' + hemi],
+                             pediatric_templates[name + '_roi2' + hemi]],
+                    'rules': [True, True],
+                    'cross_midline': False,
+                    'prob_map': prob_map_data[...,
+                                              prob_map_order.index(name + hemi)],
+                    'uid': uid}
+                uid += 1
+
+    return pediatric_bundles
```
Comment (arokem) [inline examples/plot_baby_afq.py:375]: This probably also moves into `data` or possibly into `baby.data`

**Comment by arokem** [comment]
One possible API here is that we create a new module `AFQ.baby` that includes all the baby-specific stuff. 

The question is what do we do about things that are implemented in `AFQ.api`? Should sub-class `AFQ.api.AFQ` into `AFQ.baby.api.AFQ`? We would then need to replace some of the stuff related to registration.

File: examples/plot_baby_afq.py [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+                # custom_tractography_bids_filters=None, # default
+                # ===== Registration parameters ====
+                # b0_threshold=50, # default
+                # min_bval=None, # default
+                min_bval=1000,  # override
+                # max_bval=None, # default
+                max_bval=1000,  # override
+                reg_template=pediatric_templates['UNCNeo-withCerebellum-for-babyAFQ'],
+                # reg_subject="power_map", # default
+                reg_subject="b0",  # override
```
Comment (arokem) [inline examples/plot_baby_afq.py:428]: Both of these (`reg_template` and `reg_subject`) should be skull-stripped using a brain-mask (for pilot data, we have that, for dHCP data it also comes with the anatomical pipeline derivatives). We should over-ride the pyAFQ default brain-mask (which is generated using the `median_otsu` algorithm) with the brain-mask that is provided with the respective datasets.

File: examples/plot_baby_afq.py [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+                # brain_mask=B0Mask(), # default
+                brain_mask=MaskFile("brainmask",
+                                    {"scope": "dHCP neonatal dMRI pipeline"}),
```
Comment (arokem) [inline examples/plot_baby_afq.py:431]: Yep - this is it!

Check if this is a binary mask or a continuous probability mask, and then threshold if needed (there is a class for that!).

File: examples/plot_baby_afq.py [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+                bids_path=dhcp_home,
+                dmriprep='dHCP neonatal dMRI pipeline',
+                # bids_filters={"suffix": "dwi"}, # default
+                # custom_tractography_bids_filters=None, # default
+                # ===== Registration parameters ====
+                # b0_threshold=50, # default
+                # min_bval=None, # default
+                min_bval=1000,  # override
+                # max_bval=None, # default
+                max_bval=1000,  # override
```
Comment (arokem) [inline examples/plot_baby_afq.py:425]: We might consider adding all the shells here. Proceed with caution.

Eventually, we might have some more denoising coming (through #625), so we could do DKI with multi-shell, and add some more scalars to the profiles (AWF, MK).

**Comment by arokem** [comment]
> if successful, then consider tractography using ACT within pyAFQ

Let's punt on this one for now. We'll use the tractography that is already there for the pilot data and do qsiprep/mrtrix with the dHCP data.

**Comment by arokem** [comment]
>  add segmentation end point filtering

Might not need this for this application. The MATLAB babyAFQ does not use endpoints for filtering.

**Comment by arokem** [comment]
> add segmentation end point filtering

Furthermore, the template does have a version of the AAL atlas, so if need be, we can come back to that (with the caveat that we'd need to do some massaging of the template to add some volumes that we have in the template we use with grownups).

We'll also need the ALS atlas for the definition of pAF (which, in this PR, is currently being skipped). In this case, a particular part of the AAL atlas is extracted to serve as a waypoint ROI - this is already part of the templates.

File: examples/plot_baby_afq.py (hunk 4 above)
Comment (grotheer) [reply examples/plot_baby_afq.py:431]: Just so I understand what is going on: Which registration algorithm are we using? Are we applying a brain mask in the step above or are we just checking that it is not a probability mask?

Comment (bloomdt-uw) [reply examples/plot_baby_afq.py:431]: `reg_algo=None` by default, which should use Symmetric Diffeomorphic Registration (SyN) Algorithm for registration.

In this step we are configuring pyAFQ and nothing is actually run until the `myafq.export_all()` command. According to the docs the `brain_mask` is applied before registration to a template.

File: AFQ/api.py [hunk 6]
```diff
@@ -945,6 +1070,9 @@ def _get_best_scalar(self):
         return self.scalars[0]
 
     def get_reg_template(self):
+        if isinstance(self.reg_template, nib.Nifti1Image):
```
Comment (arokem) [inline AFQ/api.py:1073]: @36000 : what do you think? Does this change make sense? Or should this go somewhere else?

File: examples/plot_baby_afq.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
+"""
+==============================
+Pediatric Profiles:
+==============================
+
+The following is an example of tractometry for pediatric bundles.
+
+TODO cite @mareike
```
Comment (arokem) [inline examples/plot_baby_afq.py:8]: https://www.biorxiv.org/content/10.1101/2021.03.29.437583v1

File: docs/source/conf.py [hunk 8]
```diff
@@ -209,6 +209,7 @@
      'examples_dirs': '../../examples',
      # path where to save gallery generated examples
      'gallery_dirs': 'auto_examples',
+     'ignore_pattern': 'plot_baby_afq.py',
```
Comment (arokem) [inline docs/source/conf.py:212]: I wonder why this is not in the `exclude_patterns` on line 82 of this file (and what the difference is between these), but maybe that's not a specific issue with this PR

File: examples/plot_baby_afq.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+import AFQ.data as afd
+
+from AFQ.definitions.mask import RoiMask, MaskFile
+
+##########################################################################
+# Get example data:
+# -----------------
+# **Diffusion dataset**
+#
+# .. note::
```
Comment (arokem) [inline examples/plot_baby_afq.py:29]: We don't actually use any dataset here, so I would say that this is all somewhat theoretical, and link to the repo that contains the code for the dHCP data analysis, to say that's what we actually do.

**Review by arokem — COMMENTED** [review]
I think that this is really close to being ready to merge! Just a few small comments here.

File: docs/source/conf.py (hunk 8 above)
Comment (bloomdt-uw) [reply docs/source/conf.py:212]: I don't recall exactly why I made this change, it was part of the original commit, so it could be OBE

**Comment by arokem** [comment]
I think this is good to go! @36000 : would you mind taking a quick look to confirm?

File: AFQ/api.py (hunk 6 above)
Comment (arokem) [reply AFQ/api.py:1073]: @36000 : do you have any thoughts here? I would like to merge this PR, but need some input from you for this particular line.

**Comment by arokem** [comment]
OK - I believe that the change alluded to [here](https://github.com/yeatmanlab/pyAFQ/pull/524#discussion_r688062629) is not harmful, so I am going to go ahead and merge this, so that we can avoid conflicts down the line.

**Pull request merged**